bin/sequencer_report

```diff
@@ -0,0 +1,265 @@
+        if result is True:
+            return "✓"
+        if result is False:
+            return "✕"
+        return "-"
+
+
+@dataclass
+class FeatureScore:
+    """Container for scored points and total points for a feature and stage"""
+
+    scored: int = 0
+    total: int = 0
+    stage: str = ""
+
+    def add(self, result, score):
+        self.total += score
+        if result == "pass":
+            self.scored += score
+
+    def has(self):
+        """Did the sequencer results have this feature & stage combination?"""
+        return self.total > 0
+
+    def passed(self):
+        """Was the feature & stage combination a pass (full marks)"""
+        return self.total > 0 and self.scored == self.total
+
+
+@dataclass(unsafe_hash=True, eq=True, order=True)
+class TestResult:
+    """Container for test result"""
+
+    bucket: str = ""
+    name: str = ""
+    result: str = ""
+    stage: str = ""
+    message: str = ""
+    score: int = DEFAULT_SCORE
+
+    def passed(self):
+        if self.esult == "pass":
+            return True
+        if self.result == "fail":
+            return False
+        return None
+
+
+@dataclass
+class SchemaResult:
+    """Container for Schema results"""
+
+    schema: str = ""
+    stage: str = ""
+    test_name: str = ""
+    result: str = ""
+
+    def passed(self):
+        if self.result == "pass":
+            return True
+        if self.result == "fail":
+            return False
+        return None
+
+
+class SequencerReport:
+    """Results of a single sequencer run"""
+
+    def __init__(self, site_path: str, device_id: str):
+        self.site_path = site_path
+        self.device_id = device_id
+        self.results_path = os.path.join(site_path, f"out/sequencer_{device_id}.json")
+
+        with open(self.results_path, encoding="utf-8") as f:
+            self.results_json = json.load(f)
+
+        self.start_time = self.results_json["start_time"]
+        self.end_time = self.results_json["status"]["timestamp"]
+        self.status_message = self.results_json["status"]["message"]
+
+        self._load_test_results()
+        self._load_schema_results()
+        self._load_device_metadata()
+
+    def _load_schema_results(self):
+        """Loads and processes schema validation results"""
+        self.schema_results = []
+        for schema, value in self.results_json["schemas"].items():
+            for test_name, result in value["sequences"].items():
+                self.schema_results.append(
+                    SchemaResult(schema, result["stage"], test_name, result["result"])
+                )
+
+        # Ensure all results include at least items from SCHEMAS set of schemas in report
+        schemas = SCHEMAS[:]
+        schemas.extend(x.schema for x in self.schema_results if x.schema not in schemas)
+
+        self.schema_results_by_schema = {
+            k: all_or_none([x.passed() for x in self.schema_results if x.schema == k])
+            for k in schemas
```

Comment:
this indent seems wrong -- shouldn't the for be +2 or +4 in?

Reply:
I think because it's part of the same list comprehension  as the `k: all_or_none` it shouldn't be indented. The linter is happy with it, and so is the black autoformatter